core/src/main/scala/org/apache/spark/executor/ProcfsMetricsGetter.scala

```diff
@@ -0,0 +1,231 @@
+      val childPidsInInt = mutable.ArrayBuffer.empty[Int]
+      def appendChildPid(s: String): Unit = {
+        if (s != "") {
+          logTrace("Found a child pid:" + s)
+          childPidsInInt += Integer.parseInt(s)
+        }
+      }
+      val stdoutThread = Utils.processStreamByLine("read stdout for pgrep",
+        process.getInputStream, appendChildPid)
+      val error = process.getErrorStream
+      var errorString = ""
+      (0 until error.available()).foreach { i =>
+        errorString += error.read()
+      }
+      val exitCode = process.waitFor()
+      stdoutThread.join()
+      // pgrep will have exit code of 1 if there are more than one child process
+      // and it will have a exit code of 2 if there is no child process
+      if (exitCode != 0 && exitCode > 2) {
+        val cmd = builder.command().toArray.mkString(" ")
+        logWarning(s"Process $cmd" +
+          s" exited with code $exitCode, with stderr:" + s"${errorString} ")
+        throw new SparkException(s"Process $cmd exited with code $exitCode")
+      }
+      childPidsInInt
+    } catch {
+      case e: Exception =>
+        logWarning("Exception when trying to compute process tree." +
+          " As a result reporting of ProcessTree metrics is stopped.", e)
+        isAvailable = false
+        mutable.ArrayBuffer.empty[Int]
+    }
+  }
+
+  def addProcfsMetricsFromOneProcess(
+      allMetrics: ProcfsMetrics,
+      pid: Int):
+      ProcfsMetrics = {
+
+    // Hadoop ProcfsBasedProcessTree class used regex and pattern matching to retrive the memory
+    // info. I tried that but found it not correct during tests, so I used normal string analysis
+    // instead. The computation of RSS and Vmem are based on proc(5):
+    // http://man7.org/linux/man-pages/man5/proc.5.html
+    try {
+      val pidDir = new File(procfsDir, pid.toString)
+      Utils.tryWithResource( new InputStreamReader(
+        new FileInputStream(
+          new File(pidDir, procfsStatFile)), Charset.forName("UTF-8"))) { fReader =>
+        Utils.tryWithResource( new BufferedReader(fReader)) { in =>
+          val procInfo = in.readLine
+          val procInfoSplit = procInfo.split(" ")
+          if (procInfoSplit != null) {
+            val vmem = procInfoSplit(22).toLong
+            val rssPages = procInfoSplit(23).toLong
+            if (procInfoSplit(1).toLowerCase(Locale.US).contains("java")) {
+              return allMetrics.copy(
+                jvmVmemTotal = allMetrics.jvmVmemTotal + vmem,
+                jvmRSSTotal = allMetrics.jvmRSSTotal + (rssPages*pageSize)
+              )
+            }
+            else if (procInfoSplit(1).toLowerCase(Locale.US).contains("python")) {
+              return allMetrics.copy(
+                pythonVmemTotal = allMetrics.pythonVmemTotal + vmem,
+                pythonRSSTotal = allMetrics.pythonRSSTotal + (rssPages*pageSize)
+              )
+            }
+              return allMetrics.copy(
```

Comment:
rather than having `return` everywhere here, more idiomatic scala to put this in an `else`, and then you can get rid of all the `return`s.